### ISSUE TITLE
Reliable service upgrade

### ISSUE THREAD

**Issue by juergstaub** [Azure/service-fabric-issues#1237]
Hi 

One of my reliable services is causing issues during upgrades. The newer version of the service contains one more IReliableDictionary which contains a new type (struct). The struct type only contains primitive types and strings. 

I get exceptions  during serialization that the code is not backward compatible. I know that the state must be forward and backward compatible, however since I only added a new collection, I hoped this would be fine. From the stacktrace I see that the issue is in deserialzing the new type/collection. 

The exception is as follows:
Below type used in Reliable Collection urn:CacheStatistic/dataStore could not be loaded. This commonly indicates that the user application is not backwards/forwards compatible. Common compatibility bugs that lead to this error are adding a new type or changing an assembly name without two phase upgrade, or removing a type. If this was caused by user's backwards/forwards compatibility bug, one way to mitigate the issue is to force the upgrade through without safety checks.\r\n\r\nSystem.Fabric.Store.TStore`5[[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089],[SomePkg.TrackService.State.CacheStatistic, SomePkg.TrackService, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null],[Microsoft.ServiceFabric.Data.Collections.DefaultComparer`1[[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], Microsoft.ServiceFabric.Data.Impl, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35],[Microsoft.ServiceFabric.Data.Collections.DefaultEqualityComparer`1[[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], Microsoft.ServiceFabric.Data.Impl, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35],[Microsoft.ServiceFabric.Data.Collections.EmptyRangePartitioner`1[[System.String, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]], Microsoft.ServiceFabric.Data.Impl, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35]], Microsoft.ServiceFabric.Data.Impl, Version=6.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35

Can I solve the issue with a two phase upgrade by adding the new type only, so that the new type is known and after deploy the full version of the service?

Thanks
Jürg

**Comment by ashishnegi** [comment]
@juergstaub did 2 phase upgrade fixed the issue ? Or something else ?

@raunakpandya Do we really need 2 phase upgrade here ?

**Comment by juergstaub** [comment]
@ashishnegi Yes it did fix it, apparently it needs the class, if it is not there it throws an exception. Maybe an upgrade without safety would have worked.

**Comment by ashishnegi** [comment]
>`adding the new type only, so that the new type is known and after deploy the full version of the service?`

@yizhang82 Do we really need a `2 phase upgrade` if are adding a new dictionary with new type ?

**Comment by juergstaub** [comment]
@ashishnegi Yes, exactly

**Comment by juergstaub** [comment]
@ashishnegi As far as I remember, it was exactly like this.

**Comment by JanardanaB** [comment]
> @ashishnegi Yes, exactly
Hi 
I am getting the same issue. added new reliable dictionary.

 can please share, how to perform 2 phase upgrade

**Comment by ashishnegi** [comment]
@JanardanaB 
1. First upgrade your app to have all the custom classes required for the dictionary. For e.g. if dictionary is `IReliableDictionary<ClassA, ClassB>`, then upgrade the app to have classes `ClassA and ClassB` but don't include the code of using it in `IReliableDictionary`.

2. Now upgrade the app again to have all your final code which has `IReliableDictionary` using `ClassA`/`ClassB`.

This makes sure that we have all classes in code packages of Primary and Secondary replicas before IReliableDictionary starts replicating the data.

**Comment by JanardanaB** [comment]
@ashishnegi ,
I am trying to reproduce the issue in sample project.
1) Created the first reliable dictionary and then published
2) Added another reliable dictionary and its supported custom classes.
 When I do the publish with different version. its saying "Upgrade completed successfully" but I see the .Net Runtime error on the event Viewer.
![image](https://user-images.githubusercontent.com/16304715/48478608-c9665f00-e82a-11e8-88aa-ebe4e3b16f9d.png)

Here I need to regenerate issue

**Comment by agopalak-cw** [comment]
@ashishnegi / @raunakpandya  - I have a similar situation, we had to add a new property to the existing reliable dictionary and upgrades throws the same error as above, "Below type used in Reliable Collection could not be loaded. This commonly indicates that the user application is not backwards/forwards compatible. Common compatibility bugs that lead to this error are adding a new type or changing an assembly name without two phase upgrade, or removing a type"

Scenario :

We have this object which is already exists in our stateful service 
public class LocationInfo { string Name, string Description, string Status }
We had to add an Id property to LocationInfo so that it becomes  LocationInfo { string Id,  string Name, string Description, string Status }

But it throws the above error,

So is there a documentation for how to make this backward compatible with Reliable Dictionary or the only solution is 2 phase deployment?

**Comment by raunakpandya** [comment]
I believe we use the data contract serializer so anything backward/forward compatible wrt the same should work. @preethasubbarayalu can you comment?

**Comment by agopalak-cw** [comment]
@raunkpandya /  @preethasubbarayalu - Did you mean that this is being handled internally in Stateful services, and we do not want to do anything from a consumer application of Stateful services? Could you please clarify?

**Comment by agopalak-cw** [comment]
More specifically my question is my object which is being already used by the ReliableDictionary in our Stateful services doesn't have any DataContract/DataMember contracts defined.  
1. If I am adding DataContract/DataMember attributes to the existing class along with adding the new property, this issue can be fixed?

**Comment by zuhairp** [comment]
If you use a custom key or value type and do not specify a custom serializer, then reliable collections will default to using DataContract serializer. If you do not declare DataContract/DataMember attributes, then by default, DataContract serializer will not be backward/forward compatible. I believe the issue can be fixed if you add those attributes.

**Comment by agopalak-cw** [comment]
Hi @zuhairp we have tried that, but still application is stuck in upgrade domain. We have a local repor for this scenario, steps as follows.

Step 1
-------------
Deploy Stateful application version with ReliableDictionary object <Location> without any DataContracts defined (This version is in our production clusters right now) locally. For Ex ,
public class Location { string Name, DateTime CreatedOn, bool Status etc. }
Result :- Successfully deployed application to Local Cluster
Step 2
-----------------
Add DataContract and MemberContract to the above LocationClass
[DataContract]
    public class Location
    {
        [DataMember]
        [JsonProperty(PropertyName = "id")]
        public string Id { get; set; }

        [DataMember]
        [JsonProperty(PropertyName = "name")]
        public string Name{ get; set; }

        [DataMember]
        [JsonProperty(PropertyName = "status")]
        public string Status{ get; set; }
    }
Upgrade the application with the newly added property and the DataContract and DataMember attributes. Deploy application locally using AutoUpgrade from Visual Studio.
Result :
Upgrade is not happening with application being in RollingForwardInProgress. for a very long time.

**Comment by zuhairp** [comment]
Can you try the following?
1. Deploy app without any DataContracts defined
2. Deploy app with DataContracts added to existing fields (don't add any new fields)
3. Deploy app with new DataMember field

**Comment by agopalak-cw** [comment]
I have tried that first, but when you add the DataMemeber/DataContract attribute to that class without adding any new property itself, the application wont upgrade. with the status

Waiting for upgrade ...
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains.
Upgrade Progress: Completed 0/1 upgrade domains. 
...................

**Comment by zuhairp** [comment]
Assuming you started out with:
```C#
class Location 
{
    public string Name { get; set; }
    public string Status { get; set; }
}
```

can you try changing it like so?
```C#
using System.Runtime.Serialization;

[Serializable] // May or may not be needed
class Location 
{
    public string Name { get; set; }
    public string Status { get; set; }
    
    [OptionalField]
    public string Id; // Note: this is a field, not a property
}
```
